addons/webhook/certificates.go

```diff
@@ -0,0 +1,203 @@
+		NotAfter:              tls.ValidFrom.Add(tls.MaxAge),
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+	cert, err := x509.CreateCertificate(cryptorand.Reader, &caTemplate, &caTemplate, &tls.ca_key_der.PublicKey, tls.ca_key_der)
+	if err != nil {
+		return nil, err
+	}
+
+	tls.ca_cert_der = &cert
+
+	//create server rsa key
+	tls.server_key_der, err = rsa.GenerateKey(cryptorand.Reader, 4096)
+	if err != nil {
+		return nil, err
+	}
+
+	//Create and sign server certificate
+	serverTemlate := x509.Certificate{
+		DNSNames:     tls.DnsNames,
+		SerialNumber: big.NewInt(2),
+		Subject: pkix.Name{
+			CommonName: fmt.Sprintf("%s@%d", tls.Hostname, time.Now().Unix()),
+		},
+		NotBefore: tls.ValidFrom,
+		NotAfter:  tls.ValidFrom.Add(tls.MaxAge),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+	cacert, err := x509.ParseCertificate(*tls.ca_cert_der)
+	if err != nil {
+		return nil, err
+	}
+	cert, err = x509.CreateCertificate(cryptorand.Reader, &serverTemlate, cacert, &tls.server_key_der.PublicKey, tls.ca_key_der)
+	if err != nil {
+		return nil, err
+	}
+	tls.server_cert_der = &cert
+	return &tls, nil
+}
+
+func (tls *TLSCredentials) WriteFiles() error {
+	//both ca and server certificates
+	certBuffer := bytes.Buffer{}
+	if err := pem.Encode(&certBuffer, &pem.Block{Type: CertificateBlockType, Bytes: *tls.server_cert_der}); err != nil {
+		return err
+	}
+	if err := pem.Encode(&certBuffer, &pem.Block{Type: CertificateBlockType, Bytes: *tls.ca_cert_der}); err != nil {
+		return err
+	}
+	// Server key
+	keyBuffer := bytes.Buffer{}
+	if err := pem.Encode(&keyBuffer, &pem.Block{Type: keyutil.RSAPrivateKeyBlockType, Bytes: x509.MarshalPKCS1PrivateKey(tls.server_key_der)}); err != nil {
+		return err
+	}
+
+	if err := certutil.WriteCert(certfile_path(), certBuffer.Bytes()); err != nil {
+		return err
+	}
+	if err := keyutil.WriteKey(keyfile_path(), keyBuffer.Bytes()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// from controller-rutime documentation at https://github.com/kubernetes-sigs/controller-runtime/blob/3dafc3c844685e4f672d9e07b88b68ae3f14a842/pkg/config/v1alpha1/types.go#L141
+// CertDir is the directory that contains the server key and certificate.
+// if not set, webhook server would look up the server key and certificate in
+// {TempDir}/k8s-webhook-server/serving-certs. The server key and certificate
+// must be named tls.key and tls.crt, respectively.
+
+//We need to come up with the functions below but for now we can use the assumed defaults
+func certfile_path() string {
+	return "/tmp/k8s-webhook-server/serving-certs/tsl.crt"
+}
+
+func keyfile_path() string {
+	return "/tmp/k8s-webhook-server/serving-certs/tsl.key"
+}
```

Comment:
This are using the default from the documentation but the paths can actually be pulled from the manager. 
Probably better to add the paths as an argument to NewTLSCredentials.